File: my_cpp_package/src/action_client_node.cpp
#include <chrono>
#include <thread>

#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "my_msgs/action/move_robot.hpp"

/**
 * @author: Mehmet Kahraman
 * @date: 09.10.2023
 * @about: Action client node
 **/


using namespace std::placeholders;
using MyActionMoveRobot = my_msgs::action::MoveRobot;
using GoalHandleMoveRobot = rclcpp_action::ClientGoalHandle<MyActionMoveRobot>;

class MyRosClass : public rclcpp::Node
{   

    public:
        
        MyRosClass() : Node("action_client_node")
        {   
            std::string action_topic = "/my_action_topic";
            
            action_client = rclcpp_action::create_client<MyActionMoveRobot>(this, action_topic);

            bool is_action_server_ready = action_client->wait_for_action_server(std::chrono::milliseconds(200));

            while (rclcpp::ok()) {
                if (is_action_server_ready == false) {
                    RCLCPP_WARN(this->get_logger(), "Waiting for %s action server ...", service_topic.c_str());
                }
                else {
                    RCLCPP_INFO(this->get_logger(), "Connected to action server node.");
                    break;
                }
                is_action_server_ready = action_client->wait_for_action_server(std::chrono::milliseconds(200));
            }

            std::this_thread::sleep_for(std::chrono::seconds(1));
            RCLCPP_INFO(this->get_logger(), "Action client node is ready.");

            execute_once();
        }

    private:

        std::string service_topic;
        rclcpp_action::Client<MyActionMoveRobot>::SharedPtr action_client;

        void send_goal(std::vector<double> pos_vector)
        {
            RCLCPP_INFO(this->get_logger(), "Sending action goal...");
            
            auto goal_msg = MyActionMoveRobot::Goal();
            goal_msg.target_position = pos_vector;

            auto send_goal_options = rclcpp_action::Client<MyActionMoveRobot>::SendGoalOptions();
            send_goal_options.goal_response_callback = std::bind(&MyRosClass::goal_response_callback, this, _1);
            send_goal_options.feedback_callback = std::bind(&MyRosClass::feedback_callback, this, _1, _2);
            send_goal_options.result_callback = std::bind(&MyRosClass::result_callback, this, _1);
            action_client->async_send_goal(goal_msg, send_goal_options);
        }

        void goal_response_callback(const GoalHandleMoveRobot::SharedPtr & goal_handle)
        {
            if (!goal_handle) 
            {
                RCLCPP_ERROR(this->get_logger(), "Goal was rejected by action server!");
            } 
            else {
                RCLCPP_INFO(this->get_logger(), "Goal accepted by action server, waiting for result.");
            }
        }

        void feedback_callback(
            GoalHandleMoveRobot::SharedPtr, 
            const std::shared_ptr<const MyActionMoveRobot::Feedback> feedback)
        {   
            int progress = feedback->progress;
            RCLCPP_INFO(this->get_logger(), "Feedback received, Progress: %d", progress);
        }

        void result_callback(const GoalHandleMoveRobot::WrappedResult & result)
        {
            switch (result.code) {
                case rclcpp_action::ResultCode::SUCCEEDED:
                    RCLCPP_INFO(this->get_logger(), "Action Goal successfully done.");
                    break;
                case rclcpp_action::ResultCode::ABORTED:
                    RCLCPP_WARN(this->get_logger(), "Action Goal was aborted");
                    return;
                case rclcpp_action::ResultCode::CANCELED:
                    RCLCPP_WARN(this->get_logger(), "Action Goal was canceled");
                    return;
                default:
                    RCLCPP_ERROR(this->get_logger(), "Unknown error occurred in the result of action!");
                    return;
            }
        }

        void execute_once()
        {   
            int seconds = 3;
            RCLCPP_INFO(this->get_logger(), "Action goal will be sent in %d seconds...", seconds);
            std::this_thread::sleep_for(std::chrono::seconds(seconds));

            auto target_pos_vector = std::vector<double>(6);
            for (size_t i=0; i<6; i++) {
                target_pos_vector[i] = 0.0;
            }

            send_goal(target_pos_vector);
        }
    

}; // class


int main(int argc, char** argv)
{
    rclcpp::init(argc, argv);
    auto action_client_node = std::make_shared<MyRosClass>();
    //rclcpp::spin(action_client_node);

    rclcpp::shutdown();
    return 0;
}

File: my_py_package/my_py_package/action_client_node.py
#!/usr/bin/env python3

import time
import rclpy
from rclpy.node import Node
from rclpy.action import ActionClient
from my_msgs.action import MoveRobot

"""
  @author: Mehmet Kahraman
  @date: 21.02.2024
  @about: Action client node
"""

class MyROSClass(Node):

    def __init__(self):

        super().__init__('action_client_node')

        self.action_topic = "/my_action_topic"
        self.action_client = ActionClient(self, MoveRobot, self.action_topic)
        
        self.get_logger().info('waiting for %s action server...' %self.action_topic)
        self.action_client.wait_for_server()
        self.get_logger().info("Action client is now connected to action server.")
    
        time.sleep(1)
        self.get_logger().info("Action client is ready.")

        self.execute_once()


    def execute_once(self):
        seconds = 3
        self.get_logger().info(f'Action goal will be sent in {seconds} seconds...')
        time.sleep(seconds)

        target_pos_vector = [0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
        self.send_goal(target_pos_vector)
        self.get_logger().info("Client process finished.")

    def send_goal(self, request_data):
        self.get_logger().info("Sending action goal...")
        goal_msg = MoveRobot.Goal()
        goal_msg.target_position = request_data
        self.action_client.wait_for_server()

        self.send_goal_future = self.action_client.send_goal_async(goal_msg, feedback_callback=self.feedback_callback)
        self.send_goal_future.add_done_callback(self.goal_response_callback)
    
    def goal_response_callback(self, future_result):
        goal_handle = future_result.result()
        if not goal_handle.accepted:
            self.get_logger().error('Goal was rejected by action server!')
            return
        else:
            self.get_logger().info('Goal accepted by action server, waiting for result.')
            self.get_result_future = goal_handle.get_result_async()
            self.get_result_future.add_done_callback(self.result_callback)

    def feedback_callback(self, feedback):
        progress = feedback.feedback.progress
        self.get_logger().info(f'Feedback received, Progress: {progress}')

    def result_callback(self, future_result):
        result = future_result.result()
        if result.result.finish_success == True:
            self.get_logger().info('Action goal successfully done.')
        else:
            self.get_logger().error('Action goal resulted with FALSE!')


def main():
    rclpy.init()
    node = MyROSClass()
    #rclpy.spin(node)

    Node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()
